numpy/random/mtrand/mtrand.pyx

```diff
@@ -620,13 +620,63 @@ def _rand_bool(low, high, size, rngstate):
         return array
 
 
-def _rand_int8(low, high, size, rngstate):
+def _rand_bool(low, high, size, rngstate):
     """
-    _rand_int8(low, high, size, rngstate)
+    _rand_bool(low, high, size, rngstate)
 
     See `_rand_int32` for documentation, only the return type changes.
 
     """
+    cdef npy_bool off, rng, buf
+    cdef ndarray array "arrayObject"
+    cdef npy_bool *array_data
+    cdef npy_bool *high_data
+    cdef npy_bool *low_data
+    cdef broadcast multi
+    cdef npy_intp i
+
+    cdef ndarray lo = <ndarray>low
+    cdef ndarray hi = <ndarray>high
+    cdef rk_state *state = <rk_state *>NpyCapsule_AsVoidPtr(rngstate)
+
+    if size is None:
+        multi = <broadcast>PyArray_MultiIterNew(2, <void *>lo, <void *>hi)
+        array = <ndarray>PyArray_SimpleNew(multi.nd, multi.dimensions, NPY_BOOL)
+        array_data = <npy_bool *>PyArray_DATA(array)
+        with nogil:
+            for i from 0 <= i < multi.size:
+                low_data = <npy_bool *>PyArray_MultiIter_DATA(multi, 0)
+                high_data = <npy_bool *>PyArray_MultiIter_DATA(multi, 1)
+                rng = <npy_bool>(high_data[0] - low_data[0])
+                off = <npy_bool>(low_data[0])
+
+                rk_random_bool(off, rng, 1, &buf, state)
+                array_data[i] = buf
+    else:
+        array = <ndarray>np.empty(size, np.bool_)
+        array_data = <npy_bool *>PyArray_DATA(array)
+        multi = <broadcast>PyArray_MultiIterNew(3, <void *>array, <void *>lo, <void *>hi)
+        if (multi.size != PyArray_SIZE(array)):
+            raise ValueError("size is not compatible with inputs")
+        with nogil:
```

Comment:
I think you can avoid duplicating this whole block of code, and put it after the `if`-`else` block, simply by making `array` the third, instead of the first, argument to `PyArray_MultiIterNew`.

If we modified, which we should, `PyArray_MultiIterNew` to also take multi-iters in, like `np.broadcast` does, it may even be possible to make this a little nicer, but that's a whole different story...